pymmcore_plus/core/_mmcore_plus.py

```diff
@@ -45,6 +45,40 @@ def __repr__(self) -> str:
 
     # Re-implemented methods from the CMMCore API
 
+    def setProperty(
+        self, label: str, propName: str, propValue: Union[bool, float, int, str]
+    ) -> None:
+        """setProperty with more reliable event emission.
+
+        As stated by Nico: "Callbacks are mainly used to give devices the opportunity to
+        signal back to the UI."
+        https://forum.image.sc/t/micromanager-events-core-events-not-coming-through/53014/2
+
+        Because it's left to the device adapter to emit a signal, in many cases uses
+        `setProperty()` will NOT lead to a new `propertyChanged` event getting emitted.
+        But that makes it hard to create listeners (i.e. in the gui or elsewhere).
+
+        While this method override cannot completely solve that problem (core-internal
+        changes will still lack an associated event emission in many cases), it can at
+        least guarantee that if we use `CMMCorePlus.setProperty` to change the property,
+        then a `propertyChanged` event will be emitted if the value did indeed change.
+
+        Parameters
+        ----------
+        label : str
+            device label
+        propName : str
+            property name
+        propValue : Union[bool, float, int, str]
+            new value
+        """
+        before = super().getProperty(label, propName)
+        with self.events.propertyChanged.blocked():  # block the native event.
+            super().setProperty(label, propName, propValue)
+        after = super().getProperty(label, propName)

```

Comment:
It's possible that this won't block the native event as I think there may be some async/threading happening where the property is set, a command goes out to the device (e.g. over serial) and some time later the device physically switches and emits a signal, and I don't think that mmcore always blocks on that. or at least once it goes into the device adapter I think all bets may be off as to what happens.

Though that's an edge case that liveable with

Reply:
it _won't_ block the native event... but we don't really want to block it unless the native event was triggered specifically by this method.  That's why this pattern is nice

Reply:
i guess you're worried that we might actually get two signals?

Reply:
Yeah you'd end up with the same signal twice which is probably fine. Although I guess that there is also potential for badness if there is any tight timing. e.g. imagine that there is a callback on a slow hardware switch that results in some fast action, then since we're emitting a signal potentially before the hardware has finished the callback would be triggered prematurely. Although I imagine that `core.watiForSystem` or `waitForDevice` would protect against this.